Fix redundant reordering of const-weight for convolution node in dynamic model

usstq opened the pull request:
### Details:
 - Make debug log on SupportedPrimitiveDescriptor prettier
 - Remove WA on 86948 (the WA which stops trying brgconv in dynamic mode)
 - Copy weight reordering design from fullyconnect node into conv:
    -  add constWeight into ConvKey 
    -  skip creating weight reorder for constWeight case in ConvolutionExecutor
    - copy prepareWeightMemory from fullyconnect node to reorder const weights
    - copy initOptimalPrimitiveDescriptor from fullyconnect node to avoid redundant reordering of const weights at graph-create time (the result of these reordering was waste of time and memory in new design)

### Tickets:
 - 86948

maxnick commented on src/plugins/intel_cpu/src/nodes/conv.cpp:
```diff
@@ -1436,6 +1443,12 @@ void Convolution::prepareParams() {
         primArgs[DNNL_ARG_DST] = dstMemPtr->GetPrimitive();
 
+        if (key.constWeight) {
+            if (!prevExecPtr || prevExecPtr->getWeightDesc() != execPtr->getWeightDesc()) {
+                primArgs[DNNL_ARG_WEIGHTS] = prepareWeightMemory(DnnlExtensionUtils::makeDescriptor(execPtr->getWeightDesc()))->GetPrimitive();
+            }
```
Is not it more reliable to do the following:
```suggestion
            auto itr = primArgs.find(DNNL_ARG_WEIGHTS);
            if (itr == primArgs.end() || itr.second.get_desc() != execPtr->getWeightDesc()) {
                primArgs[DNNL_ARG_WEIGHTS] = prepareWeightMemory(DnnlExtensionUtils::makeDescriptor(execPtr->getWeightDesc()))->GetPrimitive();
            }
```
Also that would be great to extract this procedure to some common part as well, simply to reduce the number of possible errors in implementation.

usstq commented on src/plugins/intel_cpu/src/nodes/conv.cpp:
```diff
@@ -1287,7 +1290,7 @@ void Convolution::prepareParams() {
 
     DnnlMemoryDescCPtr inMemoryDesc = srcMemPtr->GetDescWithType<DnnlMemoryDesc>();
-    DnnlMemoryDescCPtr weightMemoryDesc = wghMemPtr->GetDescWithType<DnnlMemoryDesc>();
+    DnnlMemoryDescCPtr weightMemoryDesc = MemoryDescUtils::convertToDnnlMemoryDesc(selectedWeightDesc);//wghMemPtr->GetDescWithType<DnnlMemoryDesc>();
```
That's a very good idea, I will refactor my PR according to your suggestion!

usstq commented:
@maxnick I made a few design changes in this PR:

## Move weight-reordering from load-network to first execution
 as you suggested, conv/fullyconnect CPU node now report weight memory descriptor compatible with (identical to) what specified by their corresponding ngraph OP's definition, instead of reporting blocked format queried from oneDNN primitive, so no reorder node will be inserted for weight edges in the CPU graph.

when creating the final primitive for execution, use weight memory descriptor with format tag of "any" instead of a specific layout, so here we are actually delaying the querying of actual weight layout from `initSupportedPrimitiveDescriptors()` stage to `prepareParam()` stage, because the weight-layout in former stage might be based on dummy shape, while in later stage, it's based on actual input shape and more accurate (due to that brgemm kernels may requiring weight to be of different blocking format for different input shapes).

in `prepareParam()` stage, even with weight format tag changed into `any`, creating primitive with specific combination of layout & implType may still fail, due to that some kernel checks input shapes in their initialization to determine whether the shape is suitable for its implementation. (for example, some AMX based kernel may fail to initialize itself when input shape is too small and which cannot justify a good enough HW efficiency), thus the final fallback is kept which further relaxed src/dst layout as `any` and removed implType restriction.

with these changes, weight reordering has to be done at or after `prepareParam()` stage, and it should be done in a way similar to what has been implemented in FullyConnect node: 
  - the weight reordering should be done only once if it's constant in natural (and once per execution if not constant)
  - the weight reordering result should be cached globally (in per NUMA node weight-cache) and locally(in per Node `privateWeightCache`), due to the fact that NUMA node weight-cache holds weak-references only, so the local cache is always needed if we want to save the cost of weight layout changes back-and-forth caused by input shape changes.
  - in general we cannot assume the weight data is ready at  `prepareParam()` stage even when it's constant in natural, because there may be a constant folding sub-graph between the const-Input node and conv/fullyconnect node (I'm not sure how they survived const-folding stages, but sometime they do), and these constant folding sub-graph haven't been executed at `prepareParam()` stage, so a better timing would be at `Node::execute()` stage, where all inputs are ready, including the constant one - the weight, so in this PR, I made a logic to do so in `DnnlExecutor::exec()`.

## Refactor DnnlExecutor

In current design, `DnnlExecutor` has some non-optimal behaviors:
  -  the reordering of constant weight is done on each execution which is the primary reason we started this PR;
   - input/output reorder primitives are not fetched from `ParamsCache`, so the creation cost might be not optimal.
   - the temporary memory used for reordering input/output was allocated/freed at each execution inside `DnnlExecutor::exec()`, which introduced additional runtime cost that can be optimized by allocating/re-allocating them in `prepareParam()` stage, where the required memory descriptor for them can be determined.
  - current `DnnlExecutor` only encapsulates primitive, not Prim Args, which limited what `DnnlExecutor` can do, thus all user of this class is deriving from it to extend functionality by making logics suits their own needs, and this would create redundant code, especially when we want to enhance `DnnlExecutor`.

Thus in this PR, I did a change in the design:
 - eliminate the needs for deriving from `DnnlExecutor`, by enhancing `DnnlExecutor`'s capability,  so the common code is more concentrated inside `DnnlExecutor` and API is more easier to understand and re-used.
 - incorporate prim args into `DnnlExecutor`, since it contains references to memory object, we can allocate temporary memories once and keep it in prim args to control their life-cycle.
 - add more data structures to express the execution flow, besides the input/output reorders.
 - added an API `DnnlExecutor::setArg()` to automatically determine the kind of processing needs to be done for each args, for example, runtime reordering or constant-input reordering....

## Reducing granularity of `ParamsCache` 

In current design, for nodes using `DnnlExecutor` instead of simple primitive, the `builder` callback returns an executor instead of simple primitive, this means for these nodes, the entries inside `ParamsCache` is executor rather than primitive, this has few drawbacks:
   - executor is private to each type node, and is almost opaque to other nodes, so such kind of cache entry prevents the possibility of re-using across different type of CPU node, and we have found 1 example requirement for that: Fullyconnected node can be based on brgconv1x1 primitive.
   - executor is composite of conv/ip primitive together with reorder primitive, so few different executors may actually could have share some component primitive but failed to do so due to the cache granularity.
  
so in this PR, I'm trying to fix this by reducing the granularity of  `ParamsCache` to the minimum - oneDNN primitive level (only for few important oneDNN nodes so far), and allow them to be shared across different types of CPU node, so in fullyconnect node, when it thinks brgconv1x1 is better, it can get_or_create exactly an convolution primitive from `ParamsCache`, with the exactly same type of Key that Conv CPU node would use for creating primitive, so cache sharing between them is possible. 
   
also since now we construct `DnnlExecutor` outside of `builder` function, we have to construct them at each `prepareParam()` call when input shape changes, but since all time-consuming primitives that compose the executor is being cached, I think  it should be more efficient (I assume the composing of primitives together into an executor is light-weighted).

## Canonicalizing input/output memory desc
there is a common issue of memory descriptor mismatching between CPU node & oneDNN primitive, data-type or format-tag mismatching has been handled by reordering in `DnnlExecutor`. but there is still a dimension definition mismatch, for example:
 - when 4D weight of shape [a,b,c,d] is connected to a Deconv CPU node, following ngraph's `ConvolutionBackpropData`'s convention, it assumes dimensions are presented in shape in order of (IC,OC,H,W), but when constructing a `deconvolution_forward` primitive, it requires dimensions are presented in desc in (OC,IC,H,W) order.

So when deciding if two layouts are compatible in-order to known whether a reorder is required, both descriptors should be represented in a `canonical` dimension order, so that layout/format-tag can have consistent meaning for both of them, there are 2 methods provided by oneDNN's `dnnl::memory::desc`  class to change the representation of dimension orders w/o actually change physical content of memory: `reshape()` and `permute_axes`, and they are used to `canonicalize` descriptor when passing a memory object from CPU node into a oneDNN primitive.

This canonicalization has to be done on each inference, even when input-shape is static or unchanged, due to optimizations done by `InferRequestBase::changeDefaultPtr()` can change the underlying handle(data pointer) of input memory on per-inference basis, and canonicalized memory object's underlying handle has to also be changed accordingly, and we choose to let `DnnlExecutor` to handle it by introducing another data structure recording all such needs.

Above are the major changes of this PR, please comment & review, Thanks!

usstq commented:
PS: I'm still working on rnn CPU node, current implementation using `internalBlob` should be also replaced, because:
 - it cannot sure that the final exact layout really is `dnnl_ldigo` at the `getSupportedDescriptors()` stage, especially when brgemm based kernel is chosen, the weight will be further reordered inside `prepareParams()`, this is not 100% safe, especially when there is const-folding subgraph between the weight const node and RNN node.
 - although it copied & changed gate order in `getSupportedDescriptors()`, it still reports `bac` as layout in its weight input port descriptor, this inserted an redundant reorder node on it's Weight/RecurrentWeight input ports.
 - copying original weight into `internalBlob` (with gate order changed and layout changed from `dgoi` into `ldigo`) and then reorder them later by calling `prepareMemory()` was a tricky method because all information contained in  execution Graph dump becomes confusing and misleading, due to this tricky design.
 - `internalBlob` is per node instance, it's not shared across multi streams using per NUMA weight cache.

So my plan is to:
 - add a pass in CPU graph optimizer to introduce const-folding subgraph for "gate" reordering, using `Gather` node for example, but keep the raw ngraph layout as `dgoi` because I found brgemm based rnn primitive actually prefer this layout instead of `ldigo`, in this way :
    -  we can also share this gate-reordered weight across multiple streams, to reduce memory footprint.
    - this processing steps will be explicitly shown in execution graph, to let future maintainer understand it easily
    - we can remove `internalBlob` & `prepareMemory()`  after this change
 - let RNN node read the result from that gate reordering sub-graph at first execution time and do the rest reorderings if needed, with the help of `DnnlExecutor`

usstq commented on src/plugins/intel_cpu/src/nodes/common/dnnl_executor.h:
```diff
@@ -8,41 +8,117 @@
+            : context(context),
+              name(name) {}
+
+        void reset(dnnl::primitive p, dnnl::primitive_desc_base prim_desc);
```
@maksimvlasov OK, I will push my final changes to remote (just for keeping the record, no review is needed) and continues after discussions

usstq commented on src/plugins/intel_cpu/src/nodes/conv.cpp:
```diff
@@ -1119,6 +1058,14 @@ bool Convolution::isPossibleToSkipInitConfig(DnnlDesriptor &desc) const {
 std::shared_ptr<MemoryDesc> Convolution::getSrcMemDesc(dnnl::primitive_desc_iterator &primitive_desc_it, size_t idx) {
     auto desc = idx > 0 ? primitive_desc_it.weights_desc(idx - 1) : primitive_desc_it.src_desc(idx);
+
+    if (idx == 1 && constWeightMemPtr) {
+        // original weight memory is const node, we report same descriptor so no extra reorder
+        // will be inserted by graph, so later prepareParam() can prepare Weight from original
+        // const weight node
+        return constWeightMemPtr->getDescPtr();
+    }
```
`constWeightMemPtr` was retrieved from the constant node rather than edge, so it's not `nullptr`, but you are right, we should return `plain` descriptor in all cases which is more consistent with deconv/fullyconnect's design

usstq commented on src/plugins/intel_cpu/src/nodes/common/dnnl_executor.h:
```diff
@@ -8,41 +8,117 @@
+                  handle(nullptr) {}
+        };
+
+        struct ConstFolding {
```
if you think `DnnlExecutor` as a tiny subgraph, then constant weight reordering is indeed a kind of const-folding done at the first inference stage

usstq commented on src/plugins/intel_cpu/src/nodes/common/dnnl_executor.cpp:
```diff
@@ -4,85 +4,326 @@
+    if (p_canonical_desc) {
+        cf.canonical_desc = *p_canonical_desc;
+    }
+    cf.internal_mem = dnnl::memory(expectedWeightDesc->getDnnlDesc(), context->getEngine(), nullptr);
```
OK, Great!

usstq commented on src/plugins/intel_cpu/src/nodes/conv.cpp:
```diff
@@ -375,6 +308,12 @@ void Convolution::getSupportedDescriptors() {
     withBiases = getOriginalInputsNumber() == 3;
 
+    // only here we can retrieve orginal weight memory, so later we can report weight descriptor based on this
+    auto constWeightNode = std::dynamic_pointer_cast<Input>(getParentEdgeAt(1)->getParent());
+    if (constWeightNode && constWeightNode->getType() == Type::Input && constWeightNode->isConstant()) {
+        constWeightMemPtr = constWeightNode->getMemoryPtr();
+    }
```
Yes, agree

usstq commented on src/plugins/intel_cpu/src/graph_context.h:
```diff
@@ -64,6 +65,58 @@ class GraphContext {
+                                                                      const std::vector<ptrdiff_t>& paddingL,
+                                                                      const std::vector<ptrdiff_t>& paddingR,
+                                                                      const dnnl::primitive_attr& attr,
+                                                                      const impl_desc_type& implType) const;
+
+    std::pair<dnnl::primitive, dnnl::primitive_desc_base> getConvBackPrim(const dnnl::memory::desc& src,
+                                                                          const dnnl::memory::desc& weight,
+                                                                          const dnnl::memory::desc& dst,
+                                                                          const std::vector<ptrdiff_t>& stride,
+                                                                          const std::vector<ptrdiff_t>& dilation,
+                                                                          const std::vector<ptrdiff_t>& paddingL,
+                                                                          const std::vector<ptrdiff_t>& paddingR,
+                                                                          const dnnl::primitive_attr& attr,
+                                                                          const impl_desc_type& implType) const;
+
+    std::pair<dnnl::primitive, dnnl::primitive_desc_base> getDeconvPrim(const dnnl::memory::desc& src,
+                                                                        const dnnl::memory::desc& weight,
+                                                                        const dnnl::memory::desc& bias,
+                                                                        const dnnl::memory::desc& dst,
+                                                                        const std::vector<ptrdiff_t>& stride,
+                                                                        const std::vector<ptrdiff_t>& dilation,
+                                                                        const std::vector<ptrdiff_t>& paddingL,
+                                                                        const std::vector<ptrdiff_t>& paddingR,
+                                                                        const dnnl::primitive_attr& attr,
+                                                                        const impl_desc_type& implType) const;
+
+    std::pair<dnnl::primitive, dnnl::primitive_desc_base> getInnerProductPrim(const dnnl::memory::desc& src,
+                                                                              const dnnl::memory::desc& weight,
+                                                                              const dnnl::memory::desc& bias,
+                                                                              const dnnl::memory::desc& dst,
+                                                                              const dnnl::primitive_attr& attr,
+                                                                              const impl_desc_type& implType) const;
+
+    std::pair<dnnl::primitive, dnnl::primitive_desc_base> getMatMulPrim(const dnnl::memory::desc& src,
+                                                                        const dnnl::memory::desc& weight,
+                                                                        const dnnl::memory::desc& bias,
+                                                                        const dnnl::memory::desc& dst,
+                                                                        const dnnl::primitive_attr& attr,
+                                                                        const impl_desc_type& implType) const;
+
```
Actually I suspect such kind of WA will be more and more, considering these WA should also work in dynamic shape cases, the best way is adding them in CPU node, so there maybe more 1vN mappings between CPU node and oneDNN primitives

usstq replied:
for example, anothe PR is doing this WA in ngraph transformation, but only limited to static shape case : #15019

akladiev commented:
This PR will be closed in 2 weeks in case of no activity.

akladiev commented:
This PR was closed because it has been stalled for 2 week with no activity.
